adapters/net_http/lib/opentelemetry/adapters/net/http/patches/instrumentation.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Adapters
+    module Net
+      module HTTP
+        module Patches
+          # Module to prepend to Net::HTTP for instrumentation
+          module Instrumentation
+            HTTP_METHODS_TO_SPAN_NAMES = Hash.new { |h, k| h[k] = "HTTP #{k}" }
+            USE_SSL_TO_SCHEME = { false => 'http', true => 'https' }.freeze
+
+            def request(req, body = nil, &block)
+              # Do not trace recursive call for starting the connection
+              return super(req, body, &block) unless started?
+
+              tracer.in_span(
+                HTTP_METHODS_TO_SPAN_NAMES[req.method],
+                attributes: {
+                  'component' => 'http',
+                  'http.method' => req.method,
+                  'http.scheme' => USE_SSL_TO_SCHEME[use_ssl?],
+                  'http.target' => req.path,
+                  'peer.hostname' => @address,
+                  'peer.port' => @port
+                },
+                kind: :client
+              ) do |span|
+                OpenTelemetry.propagation.inject(req)
+
+                super(req, body, &block).tap do |response|
+                  annotate_span_with_response!(span, response)
+                end
+              end
+            end
+
+            private
+
+            def annotate_span_with_response!(span, response)
+              return unless response&.code
+
+              status_code = response.code.to_i
+
+              span.set_attribute('http.status_code', status_code)
+              span.status = OpenTelemetry::Trace::Status.http_to_status(
+                status_code
+              )
```

Comment:
Do we have any semantic conventions around identifying errors based on response codes? Or are we only flagging spans as errors if an exception is raised?

Reply:
There are not semantic conventions for errors at all yet. Our `Span#record_error` method uses a set of currently proposed conventions, with the understanding that those could change. It's currently only used for unhandled exceptions. This does bring up a good point though and that is, do we want to synthesize errors based on response codes. If yes, we should probably do this across the board in our HTTP instrumentation adapters. I'll create an issue to discuss and track this.

Reply:
https://github.com/open-telemetry/opentelemetry-ruby/issues/198